CFE-1037: Add DNSNameResolver controller package

arkadeepsen opened the pull request:
Add DNSNameResolver controller package, to be imported into cluster-dns-operator. The package will be imported in openshift/cluster-dns-operator#394.

Add DNSNameResolver operator to test the controller independently.

arkadeepsen commented:
/jira refresh

candita commented:
@arkadeepsen can you please break this down into a manageable form and review the test failures (add a Makefile)?  Maybe one PR for vendor code, let that merge first.  Or several commits you can recombine later.  No more than 1000 files per commit if possible.

arkadeepsen committed:
Initial operator code scaffoled by kubebuilder.

arkadeepsen committed:
Add dnsnameresolver controller and dnsnameresolver-crd controller. Add script to fetch the dnsnameresolver crd manifest. Update Makefile to add targets for crd, verify and clean.

arkadeepsen committed:
Update Dockerfile to copy the correct directories. Add kustomization manifest for crd directory. Update rbac role permissions for running the operator.

arkadeepsen committed:
Update version of packages to maintain parity with cluster-dns-operator repo.

arkadeepsen commented:
> @arkadeepsen can you please break this down into a manageable form and review the test failures (add a Makefile)? Maybe one PR for vendor code, let that merge first. Or several commits you can recombine later. No more than 1000 files per commit if possible.

@candita I have broken down the PR into 4 commits with the first commit containing the code scaffolded by kubebuilder. The remaining commits are smaller in size. The test failure is not due to non-existence of a Makefile, rather the current CI jobs are for the dnsnameresolver plugin which is being added in #1. These tests will fail until #1 is merged.

candita commented:
/assign @alebedev87 
/assign @Miciah

arkadeepsen committed:
Update verify-gofmt.sh to exclude any vendored file.

melvinjoseph86 commented:
/jira-refresh

alebedev87 commented on operator/cmd/main.go:
```diff
@@ -0,0 +1,176 @@
+	var enableHTTP2 bool
+	var coreDNSNamespace string
+	var coreDNSServieName string
+	var dnsNameResolverNamespace string
```
```suggestion
	var (
	      metricsAddr string
	      enableLeaderElection bool
	      probeAddr string
	      secureMetrics bool
	      enableHTTP2 bool
	      coreDNSNamespace string
	      coreDNSServieName string
	      dnsNameResolverNamespace string
	)
```

alebedev87 commented on operator/cmd/main.go:
```diff
@@ -0,0 +1,176 @@
+		// the manager stops, so would be fine to enable this option. However,
+		// if you are doing or is intended to do any operation such as perform cleanups
+		// after the manager stops then its usage might be unsafe.
+		// LeaderElectionReleaseOnCancel: true,
```
`LeaderElectionReleaseOnCancel` field is not used, the docs can be found online.

alebedev87 commented on operator/go.mod:
```diff
@@ -0,0 +1,79 @@
+module github.com/openshift/coredns-ocp-dnsnameresolver/operator
+
+go 1.19
```
`go 1.21` is mentioned in the README.md and used in Dockerfile. Also this differs from `go.mod` in the root of the repository (will you use different builder images in CI?)

arkadeepsen replied:
As the DNSNameResolver controller will be imported in cluster dns operator which uses `go 1.19` I have used that version for the operator. I'll update the README.md file.

Regarding the CI jobs, we can still use image with golang version 1.21 for running the jobs as it'll work for go 1.19 as well.

alebedev87 replied:
> As the DNSNameResolver controller will be imported in cluster dns operator which uses go 1.19 I have used that version for the operator.

Did you see any compile errors related to this? I don't think the controller uses any 1.21 constructions since it builds green with `go 1.19`.

arkadeepsen replied:
There were no compile errors related to this.

alebedev87 commented on operator/Dockerfile:
```diff
@@ -0,0 +1,33 @@
+COPY go.sum go.sum
+# cache deps before building and copying source so that we don't need to re-download as much
+# and so that source changes don't invalidate our downloaded layer
+RUN go mod download
```
Why not `COPY vendor`? `go mod download` often takes quite some time which complicates the frequent local builds.

arkadeepsen replied:
The scaffolded code didn't have the vendor directory. I added it later on. The Dockerfile was scaffolded by kubebuilder that's why it uses `go mod download`. I will change it to `COPY vendor`.

alebedev87 commented on operator/Dockerfile:
```diff
@@ -0,0 +1,33 @@
+# by leaving it empty we can ensure that the container and binary shipped on it will have the same platform.
+RUN CGO_ENABLED=0 GOOS=${TARGETOS:-linux} GOARCH=${TARGETARCH} go build -a -o manager cmd/main.go
+
+# Use distroless as minimal base image to package the manager binary
+# Refer to https://github.com/GoogleContainerTools/distroless for more details
+FROM gcr.io/distroless/static:nonroot
```
Why not UBI or RHEL? Do we target non Red Hat "audience" too?

arkadeepsen replied:
This Dockerfile is needed for upstream ovnk e2e tests. That's why this base image is used.

alebedev87 commented on operator/Makefile:
```diff
@@ -0,0 +1,216 @@
+# (i.e. docker build --platform linux/arm64). However, you must enable docker buildKit for it.
+# More info: https://docs.docker.com/develop/develop-images/build_enhancements/
+.PHONY: docker-build
+docker-build: ## Build docker image with the manager.
```
```suggestion
image-build: ## Build docker image with the manager.
```
To make things look little less vendor locked.

alebedev87 replied:
Just to finish up on this one: we now have `image-build` and `docker-push` targets.

alebedev87 commented on operator/Makefile:
```diff
@@ -0,0 +1,216 @@
+# To adequately provide solutions that are compatible with multiple platforms, you should consider using this option.
+PLATFORMS ?= linux/arm64,linux/amd64,linux/s390x,linux/ppc64le
+.PHONY: docker-buildx
+docker-buildx: ## Build and push docker image for the manager for cross-platform support
```
Do we really need this target? Do we plan to build for multiple architectures? It's big and complex if we don't see how we can use it now, let's render an already decorated Makefile simpler.

arkadeepsen replied:
I can remove this target.

alebedev87 commented on operator/Makefile:
```diff
@@ -0,0 +1,216 @@
+	rm Dockerfile.cross
+
+.PHONY: build-installer
+build-installer: manifests generate kustomize ## Generate a consolidated YAML with CRDs and deployment.
```
Same remark about this rule.

arkadeepsen replied:
It's good to have this target which will be handy if we want to test the operator in the upstream ovnk e2e tests.

alebedev87 commented on operator/hack/update-generated-crd.sh:
```diff
@@ -0,0 +1,21 @@
+
+# Can't rely on associative arrays for old Bash versions (e.g. OSX)
+install_crd \
+  "vendor/github.com/openshift/api/network/v1alpha1/0000_70_dnsnameresolver_00-techpreview.crd.yaml" \
```
The api repository changed recently. It moved the generated CRDs into a separate package: `zz_generated.crd-manifests`. Do we have a CI to catch this?

arkadeepsen replied:
We are currently using an older version for openshift/api which will match with that of cluster-dns-operator. As we are importing the DNSNameResolver controller into CDO, any mismatch in the version in the go.mod file will cause the version to change in the go.mod file of the CDO.

alebedev87 commented on operator/hack/update-generated-crd.sh:
```diff
@@ -0,0 +1,21 @@
+set -euo pipefail
+
+function install_crd {
+  if [[ -z "${SKIP_COPY+1}" ]]; then
```
Do we use `SKIP_COPY` anywhere? What does it do?

arkadeepsen replied:
It can be used when calling the `manifests` make target. If some value is assigned to `SKIP_COPY` then the copy of the CRD will be skipped.

alebedev87 replied:
The only use case I see for this is when I want to change the CRD locally to test an API change. I think that in this case I would comment `update-generated-crd.sh` in Makefile so that this leaves a trace in git and I won't forget to remove it before I commit and push. Instead of letting my environment decide whether the drift should be tolerated. This is not a big deal, it just may result into unwanted `verify` CI job failures if the local environment is set with `SKIP_COPY=on`.

arkadeepsen replied:
I am fine with either option. For your reference, I used the similar format of CDO: https://github.com/openshift/cluster-dns-operator/blob/master/hack/update-generated-crd.sh

alebedev87 commented on operator/README.md:
````diff
@@ -0,0 +1,120 @@
+make deploy IMG=<some-registry>/operator:tag
+```
+
+> **NOTE**: If you encounter RBAC errors, you may need to grant yourself cluster-admin 
+privileges or be logged in as admin.
````
```suggestion
> **NOTE**: If you encounter RBAC errors, you may need to login as cluster-admin.
```

alebedev87 commented on operator/README.md:
```diff
@@ -0,0 +1,120 @@
+privileges or be logged in as admin.
+
+**Create instances of your solution**
+You can apply the samples (examples) from the config/sample:
```
Empty line is missing after the chapter title.

arkadeepsen replied:
I'll remove this section.

alebedev87 commented on operator/README.md:
```diff
@@ -0,0 +1,120 @@
+> **NOTE**: If you encounter RBAC errors, you may need to grant yourself cluster-admin 
+privileges or be logged in as admin.
+
+**Create instances of your solution**
```
Solution?

arkadeepsen replied:
I'll remove this section.

alebedev87 commented on operator/README.md:
```diff
@@ -0,0 +1,120 @@
+To test the unmanaged DNSNameResolver controller, the operator runs another controller which
+watches the DNSNameResolver CRD and starts the DNSNameResolver controller.
+
+Kindly provide correct values for the following arguments when deploying the operator:
```
Where the user should put those values?

arkadeepsen replied:
The user should provide the arguments through the `args` field of the manager deployment.

alebedev87 replied:
We should mention this in the README, just for the clarity of the instructions.

alebedev87 commented on operator/README.md:
````diff
@@ -0,0 +1,120 @@
+make uninstall
+```
+
+**UnDeploy the controller from the cluster:**
````
```suggestion
**UnDeploy the operator from the cluster:**
```

alebedev87 commented on operator/README.md:
````diff
@@ -0,0 +1,120 @@
+make undeploy
+```
+
+## Project Distribution
````
Do we need this chapter?

arkadeepsen replied:
I can remove the chapter.

alebedev87 commented on hack/verify-gofmt.sh:
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-go_files=$( find . -name '*.go' -print )
+go_files=$( find . -name '*.go' -not -path '*/vendor/*' -print )
```
I suppose the whole operator is not needed to be verified for the go format as it has its own Makefile targets for this.
```suggestion
go_files=$( find . -name '*.go' -not -path '*/operator/*' -print )
```

alebedev87 reviewed (commented):
Part 1. Makefile, Dockerfile, README.

Do we have a CI for the operator code/test/image build? If not we need it as a per-requisite for this PR to be merged.

arkadeepsen committed:
Add verify-gofmt.sh for opertor and call the script in verify target of make.
Update verify-gofmt.sh of plugin.
Copy vendor directory instead of go mod download in the Dockerfile.
Update Makefile of operator.
Remove sections from README.md which are not required.
Comment out crd/... from config/default directory's kustomization.yaml as it will cause issues when using `make install/deploy/undeploy/uninstall`.
Update main.go.

alebedev87 commented on operator/controller/dnsnameresolver-crd/controller.go:
```diff
@@ -0,0 +1,132 @@
+		return nil, err
+	}
+	clusterNamePredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		crd := manifests.DNSNameResolverCRD().DeepCopy()
```
It seems to me that the deep copy is unnecessary, you don't change the  object.

alebedev87 commented on operator/controller/dnsnameresolver-crd/controller.go:
```diff
@@ -0,0 +1,132 @@
+		actualName := types.NamespacedName{
+			Name: o.GetName(),
+		}
+		return expectedName == actualName
```
```suggestion
		return manifests.DNSNameResolverCRD().Name == o.GetName()
```

alebedev87 commented on operator/controller/dnsnameresolver-crd/controller.go:
```diff
@@ -0,0 +1,132 @@
+
+			// Start the dependent cache.
+			go func() {
+				if err := (*cache).Start(ctx); err != nil {
```
```suggestion
				if err := cache.Start(ctx); err != nil {
```

arkadeepsen replied:
I took inspiration from cluster-ingress-operator code which gets the pointer to the `controller.Controller` interface and then deference it using `(*c)`. Maybe @Miciah can provide some insights as to why this approach was used. It might be because the controller variable in the goroutine it is made sure that the actual controller variable is used rather than a copied value of it.

https://github.com/openshift/cluster-ingress-operator/commit/05c0bcb4409fcd5c4a7b349bc1961ce61dd346f6#diff-74bd64d7a0ece33319f689d27be640577ca56ace22c49b0a07192575c5a8ffb1R104-R114

alebedev87 replied:
The `cache` variable is created for each `for` iteration so we are not suppose to have a problem with the access to a shared variable. The only thing I can think of is the difference in the receiver object passed to `Start` method. Maybe a value receiver is expected to implement the `Informers` interface.

alebedev87 replied:
For me it compiles well if `cache` and `controller` variables are of type `interface` and not a pointer to an `interface`:

```diff
$ make build
hack/update-generated-crd.sh
/tmp/coredns-ocp-dnsnameresolver/operator/bin/controller-gen-v0.14.0 rbac:roleName=manager-role crd webhook paths="./..." output:crd:artifacts:config=config/crd/bases
/tmp/coredns-ocp-dnsnameresolver/operator/bin/controller-gen-v0.14.0 object:headerFile="hack/boilerplate.go.txt" paths="./..."
go fmt ./...
go vet ./...
go build -o bin/manager cmd/main.go

$ git diff
diff --git a/operator/controller/dnsnameresolver-crd/controller.go b/operator/controller/dnsnameresolver-crd/controller.go
index 9535e93..d8ceebb 100644
--- a/operator/controller/dnsnameresolver-crd/controller.go
+++ b/operator/controller/dnsnameresolver-crd/controller.go
@@ -92,11 +92,11 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
        r.startCaches.Do(func() {
                var wg sync.WaitGroup
                for i := range r.config.DependentCaches {
-                       cache := &r.config.DependentCaches[i]
+                       cache := r.config.DependentCaches[i]
 
                        // Start the dependent cache.
                        go func() {
-                               if err := (*cache).Start(ctx); err != nil {
+                               if err := cache.Start(ctx); err != nil {
                                        controllerLog.Error(err, "cannot start cache")
                                }
                        }()
@@ -104,7 +104,7 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
                        // Wait for the dependent cache to sync.
                        wg.Add(1)
                        go func() {
-                               if started := (*cache).WaitForCacheSync(ctx); !started {
+                               if started := cache.WaitForCacheSync(ctx); !started {
                                        controllerLog.Info("failed to sync cache before starting controllers")
                                }
```

alebedev87 commented on operator/controller/dnsnameresolver-crd/controller.go:
```diff
@@ -0,0 +1,132 @@
+			}()
+
+			// Wait for the dependent cache to sync.
+			wg.Add(1)
+			go func() {
+				if started := (*cache).WaitForCacheSync(ctx); !started {
+					controllerLog.Info("failed to sync cache before starting controllers")
+				}
+				wg.Done()
```
```suggestion
                        wg.Add(1)
                        go func() {
			        defer wg.Done()
				if started := cache.WaitForCacheSync(ctx); !started {
					controllerLog.Info("failed to sync cache before starting controllers")
				}
```

Miciah replied:
@alebedev87, would that introduce a race condition?  What happens if execution reaches `wg.Wait()` before the goroutine gets a chance to run?  

Setting aside my question, was there a problem with the code as Arkadeep has it written?

alebedev87 replied:
> would that introduce a race condition? What happens if execution reaches wg.Wait() before the goroutine gets a chance to run?

Right, I didn't notice I moved `Add` too.

> was there a problem with the code as Arkadeep has it written?

Not really , just wanted to defer the `Done` function to make a little more idiomatic.

alebedev87 commented on operator/controller/dnsnameresolver-crd/controller.go:
```diff
@@ -0,0 +1,132 @@
+		}
+		// Wait for all the dependent caches to sync.
+		wg.Wait()
+	})
```
```suggestion
		wg.Wait()
		сontrollerLog.Info("dependent caches synced")
	})
```

alebedev87 commented on operator/controller/dnsnameresolver-crd/controller.go:
```diff
@@ -0,0 +1,132 @@
+
+			// Start the dependent controller.
+			go func() {
+				if err := (*controller).Start(ctx); err != nil {
```
```suggestion
				if err := controller.Start(ctx); err != nil {
```

arkadeepsen replied:
Same as https://github.com/openshift/coredns-ocp-dnsnameresolver/pull/2#discussion_r1577396696

alebedev87 commented on operator/controller/dnsnameresolver-crd/controller.go:
```diff
@@ -0,0 +1,132 @@
+				}
+			}()
+		}
+	})
```
```suggestion
		}
		сontrollerLog.Info("dependent controllers started")
	})
```

alebedev87 commented on operator/controller/dnsnameresolver-crd/controller.go:
```diff
@@ -0,0 +1,132 @@
+			// Start the dependent cache.
+			go func() {
+				if err := (*cache).Start(ctx); err != nil {
+					controllerLog.Error(err, "cannot start cache")
```
What if one of the caches fails to start (or fails to sync)? Since we don't retry in case of an error, the dependent controller will start without the cache?

arkadeepsen replied:
I will change the code so that if the caches fail to start or sync then the controller won't be started. However, as we are doing this only once across all reconciles, this process won't be retried.

alebedev87 replied:
As discussed in [this thread](https://redhat-internal.slack.com/archives/C04VCUBSU4U/p1713964224175029), let's follow up on this one in a separate PR.

alebedev87 commented on operator/controller/dnsnameresolver-crd/crds.go:
```diff
@@ -0,0 +1,127 @@
+	}
+
+	// Diff before updating because the client may mutate the object.
+	diff := cmp.Diff(current, updated, cmpopts.EquateEmpty())
```
The `Diff` function can be used in `crdChanged` function and can be returned from it. This way you will save 1 comparison.

alebedev87 commented on operator/cmd/main.go:
```diff
@@ -0,0 +1,167 @@
+		os.Exit(1)
+	}
+
+	// Set up the DNSNameResolver controller.  This controller is unmanaged by
+	// the manager; the dnsnameresolvercrd controller starts it and the
+	// caches after it creates the DNSNameResolver CRD.
```
Can you explain why it's not managed by the manager?

Miciah replied:
The reason why the controller is unmanaged is so that it doesn't get automatically started when the operator starts; we only want the controller to start if we need it.  It would be good for the comment to explain this.

alebedev87 replied:
As discussed in [this thread](https://redhat-internal.slack.com/archives/C04VCUBSU4U/p1713964224175029), let's follow up on this one in a separate PR.

alebedev87 commented on operator/controller/dnsnameresolver/controller.go:
```diff
@@ -0,0 +1,225 @@
+	corednsEndpointsSliceCache, err := cache.New(mgr.GetConfig(), cache.Options{
+		Scheme: mgr.GetScheme(),
+		DefaultNamespaces: map[string]cache.Config{
+			config.OperandNamespace: {},
```
Can't we have a single cache for both `DNSNameResolver` and `EndpointSlice` resources?

We can add 2 keys to `DefaultNamespaces` map: `config.DNSNameResolverNamespace` with default config and `config.OperandNamespace` with a label selector config .

Miciah replied:
I had a similar comment here: https://github.com/openshift/cluster-dns-operator/pull/394#discussion_r1491782222

@alebedev87, does Aurko's response to my comment answer your question?

alebedev87 commented on operator/controller/dnsnameresolver/controller.go:
```diff
@@ -0,0 +1,225 @@
+	// No reconcile requests should be generated.
+	if err := c.Watch(source.Kind(corednsEndpointsSliceCache, &discoveryv1.EndpointSlice{}),
+		handler.EnqueueRequestsFromMapFunc(func(context.Context, client.Object) []reconcile.Request {
+			return nil
```
Why not just the standard `handler.EnqueueRequestForObject`? You don't use the passed object anyway.

Miciah replied:
@alebedev87, `EnqueueRequestForObject` _would_ cause a reconciliation request.  My understanding from the comment on lines 103 and 104 is that the purpose of the `Watch` is just to make controller-runtime keep a cache of endpointslices, and the purpose of `EnqueueRequestsFromMapFunc` / `return nil`  is avoid this `Watch` from causing undesired reconciliation requests.

alebedev87 replied:
I see now. The api client managed by the controller runtime manager could be used for the same purpose. In that case the read API calls would be direct though: more accurate but more costly for the api server (however this is hard to verify).

alebedev87 commented on operator/controller/dnsnameresolver/controller.go:
```diff
@@ -0,0 +1,225 @@
+		for _, resolvedAddress := range resolvedName.ResolvedAddresses {
+			if time.Now().Before(resolvedAddress.LastLookupTime.Time.Add(
+				time.Second*time.Duration(resolvedAddress.TTLSeconds) + ipRemovalGracePeriod)) {
+				updateResolvedAddresses = append(updateResolvedAddresses, resolvedAddress)
```
Shouldn't `updated=true` be here? You add this slice to the status later.

Miciah replied:
The way I read this is that we're reconstructing the list of addresses in `status.ResolvedNames[index].ResolvedAddresses`, and if we keep every address in that list, then we've made no change, so we only set `updated = true` if we *don't* keep some address (i.e. we don't append the address to the reconstructed list).

alebedev87 commented on operator/controller/dnsnameresolver/resolver.go:
```diff
@@ -0,0 +1,377 @@
+
+// Start starts a goroutine for sending DNS lookup requests of the DNS names on expiration of the corresponding TTLs.
+func (res *Resolver) Start() {
+	var nextDNSName, deletedDNSName string
+	var nextLookupTime time.Time
+	var exists bool
+	var numIPs int
```
```suggestion
	var (
	       nextDNSName string
	       deletedDNSName string
	       nextLookupTime time.Time
	       exists bool
	       numIPs int
	 )      
```

alebedev87 replied:
Btw why are they declared here and not in the goroutine where they are used?

arkadeepsen replied:
There's no specific reason for doing so. Just wanted to define the variables upfront. However, defining them within the goroutine also would not have any effect on the overall outcome of the function.

alebedev87 commented on operator/controller/dnsnameresolver/resolver.go:
```diff
@@ -0,0 +1,377 @@
+	}
+
+	var randomIPs []string
+	r1 := rand.New(rand.NewSource(time.Now().UnixNano()))
```
This is only needed in `else` case so it can be moved there.

alebedev87 commented on operator/controller/dnsnameresolver/resolver.go:
```diff
@@ -0,0 +1,377 @@
+
+// getRandomCoreDNSPodIP returns randomly chosen CoreDNS pod IPs. The input maxIPs defines
+// the upper limit for the number of CoreDNS pod IPs returned.
+func (res *Resolver) getRandomCoreDNSPodIP(maxIPs int) ([]string, error) {
```
```suggestion
func (res *Resolver) getRandomCoreDNSPodIPs(maxIPs int) ([]string, error) {
```

alebedev87 commented on operator/controller/dnsnameresolver/resolver.go:
```diff
@@ -0,0 +1,377 @@
+
+// lookupDNSName sends a DNS lookup request to CoreDNS pod(s). The DNS lookup is performed to
+// trigger an update, if required, of the DNSNameResolver resources matching the DNS name.
+func (res *Resolver) lookupDNSName(dnsName string, numIPs int) error {
```
```suggestion
func (res *Resolver) lookupDNSNameFromCoreDNS(dnsName string, numIPs int) error {
```

alebedev87 commented on operator/controller/dnsnameresolver/controller.go:
```diff
@@ -0,0 +1,225 @@
+	// event then this will ensure that the resolver sends the resolution request
+	// for the DNS name, so that the status of the object gets updated with the
+	// corresponding IP addresses.
+	r.resolver.Add(dnsName, []ocpnetworkv1alpha1.DNSNameResolverResolvedAddress{}, matchesRegular, request.Name)
```
Should we do this `Add` if the status contains some resolved names?

arkadeepsen replied:
This `Add` call is required during the create event of the `DNSNameResolver` object. At that time the resolver needs to send the DNS lookup request for the DNS name in the object so that the CoreDNS plugin can update the status if the `DNSNameResolver` object.

alebedev87 commented on operator/controller/dnsnameresolver/resolver.go:
```diff
@@ -0,0 +1,377 @@
+
+	for dnsName, resolvedName := range res.dnsNames {
+		if !exists || resolvedName.minNextLookupTime.Before(minNextLookupTime) {
+			exists = true
+			minNextLookupTime = resolvedName.minNextLookupTime
+			dns = dnsName
+			numIPs = resolvedName.numIPs
```
Why not returning right here?
Do we really need to return `exists`? `dns` can be empty in case it doesn't exist.

Miciah replied:
If you return here, then you might not actually get the `dnsName` that has the lowest `minNextLookupTime`.  Adding a `return` here causes `TestResolver` to fail.

arkadeepsen replied:
This function is used to get the DNS name with the `minNextLookupTime` so that we can look it up in the next round. Every round we need to do this exercise to find out the DNS name with `minNextLookupTime` and then send DNS lookup request for the same.

alebedev87 commented on operator/controller/dnsnameresolver/resolver.go:
```diff
@@ -0,0 +1,377 @@
+			if !first || isBeforeMinNextLookupTime {
+				resolvedName.minNextLookupTime =
+					resolvedAddress.LastLookupTime.Time.Add(time.Second * time.Duration(resolvedAddress.TTLSeconds))
+				first = true
```
Can't you just `break` here instead of using `first` variable?

Miciah replied:
Same issue as https://github.com/openshift/coredns-ocp-dnsnameresolver/pull/2#discussion_r1576625494: ~returning~ breaking means you may not get the actual min lookup time and causes test failures.

alebedev87 commented on operator/controller/dnsnameresolver/resolver.go:
```diff
@@ -0,0 +1,377 @@
+			serverStr := net.JoinHostPort(coreDNSPodIP, defaultPort)
+			dnsClient := &dns.Client{
+				Timeout: defaultMinTTL,
+			}
```
`dnsClient` can be created only once for both record types (before `for` loop).

alebedev87 commented on operator/controller/dnsnameresolver/resolver.go:
```diff
@@ -0,0 +1,377 @@
+		dnsNames:                  make(map[string]dnsResolvedName),
+		regularObjInfo:            make(map[string]string),
+		wildcardObjInfo:           make(map[string]sets.Set[string]),
+		added:                     make(chan struct{}, 1),
+		deleted:                   make(chan string, 1),
```
Any particular reason they are buffered? Why `1`?

arkadeepsen replied:
We are passing values through the channel that's why it is buffered. As we want the channel to hold only one value at a time that's why it is `1`.

alebedev87 replied:
The idea is to not block the execution of `Resolver.Add()` and `Resolver.Delete()` until `Resolver.Start()` gets the value? If so, why not leaving a bigger leeway than just 1 element? The buffered channel of size `1` will turn into a blocking on the second call.

arkadeepsen replied:
The idea is that on every add/delete the next dns name and the next lookup time needs to be evaluated by `getNextDNSNameDetails()`. If we don't block then we may not be able to ensure that we get the correct dns name on every add/delete event of DNS names. Also the channels will be unblocked whenever we complete the call to `getNextDNSNameDetails()`.

alebedev87 replied:
> The idea is that on every add/delete the next dns name and the next lookup time needs to be evaluated by getNextDNSNameDetails(). If we don't block then we may not be able to ensure that we get the correct dns name on every add/delete event of DNS names. 

With the buffered channel of size 1, the first write call is not blocking. From what you wrote here - you need a normal (blocking) channel, not buffered one.

arkadeepsen replied:
Got it. Will remove the length from the make call for the channels.

alebedev87 commented on operator/controller/dnsnameresolver/resolver_test.go:
```diff
@@ -0,0 +1,183 @@
+	}
+
+	// Create the Resolver object.
+	res := NewResolver(nil, "")
```
The tests depend on each other? Can we make test cases isolated? Each test case can get a list of add/delete actions.

alebedev87 replied:
As discussed in [this thread](https://redhat-internal.slack.com/archives/C04VCUBSU4U/p1713964224175029), let's follow up on the testing in a separate PR.

alebedev87 commented on operator/controller/dnsnameresolver/resolver_test.go:
```diff
@@ -0,0 +1,183 @@
+			assert.Equal(t, tc.expectedOutput, exists)
+			if exists {
+				cmpOpts := cmpopts.EquateApproxTime(100 * time.Millisecond)
+				diff := cmp.Diff(tc.expectedNextLookupTime, nextLookupTime, cmpOpts)
```
Diff for 2 time objects?

arkadeepsen replied:
This is used to check if the two time objects are within 100 ms of each other.

alebedev87 replied:
> This is used to check if the two time objects are within 100 ms of each other.

Will this be seen in the diff?

arkadeepsen replied:
> > This is used to check if the two time objects are within 100 ms of each other.
> 
> Will this be seen in the diff?

Yes. An example diff:
```
time.Time(
        - 	s"2024-04-24 16:58:56.239758026 +0530 IST m=+8.007328537",
        + 	s"2024-04-24 16:58:53.239757695 +0530 IST m=+5.007328207",
         )
```

alebedev87 replied:
I see:
```
    --- FAIL: TestResolver/Add_a_regular_resolved_name_belonging_to_a_wildcard_DNSNameResolver_object (0.00s)
        resolver_test.go:178:   time.Time(
            - 	s"2024-04-24 13:43:36.20190963 +0200 CEST m=+100.015875820",
            + 	s"2024-04-24 13:42:04.201908893 +0200 CEST m=+8.015875084",
              )
```

I think it's fine, may be a little explanation message would be of help:
```go
t.Fatalf("unexpected next lookuptime (-want +got): %s\n", diff)
```

alebedev87 commented on operator/controller/dnsnameresolver/resolver_test.go:
```diff
@@ -0,0 +1,183 @@
+	numRemoved       int
+}
+
+func TestResolver(t *testing.T) {
```
The lookup and the coreDNS POD IP selection are not tested.

alebedev87 replied:
As discussed in [this thread](https://redhat-internal.slack.com/archives/C04VCUBSU4U/p1713964224175029), let's follow up on the testing in a separate PR.

alebedev87 commented on operator/config/crd/bases/network.openshift.io.openshift.io_dnsnameresolvers.yaml:
Why do you need it as a part of the config manifests? You ensure this CRD from the assets.

arkadeepsen replied:
We need it as part of the deployment config as without the CRD being deployed along with the operator, the crd controller will not get any event and won't start the dependent dnsnameresolver controller.

alebedev87 replied:
I see, since the crd controller uses a predicate to watch only `DNSNameResolver` CRD it cannot get reconciliation event with the CRD being present. But this raises another question - why do we need [the complication with unmanaged and dependent controllers](https://github.com/openshift/coredns-ocp-dnsnameresolver/pull/2#discussion_r1576572562) then? If the CRD is present the first controller can be started by the controller-runtime's manager, no?

arkadeepsen replied:
This operator is a standalone one to test the functionality upstream. In OCP we will enable the unmanaged dnsnameresolver controller when the `DNSNameResolver` feature gate is enabled in CDO. That's why we need it to be unmanaged. We are importing this controller in CDO here: https://github.com/openshift/cluster-dns-operator/pull/394/files#diff-99c0290799daf9abc6240df64063e20bfaf67b371577b67ac7eec6f4725622ffR139-R165

alebedev87 replied:
As discussed in [the thread](https://redhat-internal.slack.com/archives/C04VCUBSU4U/p1713964224175029), let's follow up on this one in a separate PR.

alebedev87 commented on operator/config/manager/manager.yaml:
```diff
@@ -0,0 +1,102 @@
+          initialDelaySeconds: 5
+          periodSeconds: 10
+        # TODO(user): Configure the resources accordingly based on the project requirements.
+        # More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
```
You did configure the resource requests/limits, this TODO can be removed.

alebedev87 commented on operator/config/manager/manager.yaml:
```diff
@@ -0,0 +1,102 @@
+        resources:
+          limits:
+            cpu: 500m
+            memory: 128Mi
```
Are you sure about such a low limit? The container will be restarted if the limit is over passed. I think you can remove it to let the kernel manage the memory pressure.

Miciah replied:
If we're considering this controller to behave more as a platform component rather than as workload, then https://github.com/openshift/enhancements/blob/master/CONVENTIONS.md#resources-and-limits applies:

> Unlike with user workloads, setting limits for cluster components is problematic for several reasons:
> 
> [...]
> 
> Therefore, cluster components SHOULD NOT be configured with resource limits.
>
> However, cluster components MUST declare resource requests for both CPU and memory.

alebedev87 commented on operator/test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,121 @@
+
+var _ = Describe("controller", Ordered, func() {
+	BeforeAll(func() {
+		By("installing prometheus operator")
+		Expect(utils.InstallPrometheusOperator()).To(Succeed())
+
+		By("installing the cert-manager")
+		Expect(utils.InstallCertManager()).To(Succeed())
```
You don't add either of them to the kustomize output. Why do you install them then?

alebedev87 replied:
As discussed in [this thread](https://redhat-internal.slack.com/archives/C04VCUBSU4U/p1713964224175029), let's follow up on the testing in a separate PR.

alebedev87 commented on operator/test/e2e/e2e_suite_test.go:
```diff
@@ -0,0 +1,32 @@
+)
+
+// Run e2e tests using the Ginkgo runner.
+func TestE2E(t *testing.T) {
```
I don't see this test to be part of [the CI](https://github.com/openshift/release/blob/fcf6ab3ef8ca0452f0333bb0cc5ac328ac6a3b9a/ci-operator/config/openshift/coredns-ocp-dnsnameresolver/openshift-coredns-ocp-dnsnameresolver-main.yaml#L18).

alebedev87 replied:
As discussed in [this thread](https://redhat-internal.slack.com/archives/C04VCUBSU4U/p1713964224175029), let's follow up on the testing in a separate PR.

alebedev87 reviewed (commented):
The resolver logic is convoluted. I think I won't be able to analyze it thoroughly in the given time limits. That's why I think the unit tests for the resolver should be extended. The rest is mostly the nitpicks.

arkadeepsen committed:
Update README.md to add info about the location to provide required arguments. Add comment explaining the need for unmanaged controller. Uodate tests for resolver. Minor changes to enablednsnameresolver controller.

alebedev87 commented:
As discussed in [this thread](https://redhat-internal.slack.com/archives/C04VCUBSU4U/p1713964224175029), let's follow up on the testing and the code simplification in separate PRs.

/lgtm

Miciah commented:
/approve

snarayan-redhat commented:
/label docs-approved

arkadeepsen commented:
@melvinjoseph86 adding the label as the changes in this PR will be imported into openshift/cluster-dns-operator#394
/label qe-approved

arkadeepsen commented:
@CFields651 adding the label as the changes in this PR will be imported into https://github.com/openshift/cluster-dns-operator/pull/394
/label px-approved
